### PR TITLE
Add OffDataBackgroundMaker

### PR DESCRIPTION
**Pull request by JouvinLea**
PR for the script of how to use the new class.
Tell me what you think! It's quite the same than the test_background_model!

### CODE REVIEW

**Comment by cdeil** [comment]
If I had started the `examples/test_backgroundmodel.py`, it would probably have been this (code not tested, just typed):
```python
from gammapy.data import DataStore
from gammapy.background import OffDataBackgroundMaker

def test_background_model():
    data_store = DataStore.from_dir('$GAMMAPY_EXTRA/datasets/hess-crab4-hd-hap-prod2/')
    out_dir = # somewhere where all temp / output files will go
    bgmaker = OffDataBackgroundMaker(data_store, out_dir)
    bgmaker.select_observations( ... options ... )
    bgmaker.group_observations( ... options ... )
    bgmaker.make_model(... options ... )
    bgmaker.save()

if __name__ == '__main__':
    test_background_model()
```

And then I'd comment in / out the method calls for the steps I want to test.
It's very similar to your test script, except it's shorter.

I'm not sure about how exactly the class should work. There's many detailed decisions to make, e.g. when to store things on disk or only as members on the `bgmaker` class. Or which options to take in the `__init__` and which in later methods ...

Actually it could be useful to have two or three such small `test_background_model` functions that represent different use cases such as 2D or 3D model, or the case where the obs list is computed via a method or given as input from the outside by the caller (e.g. by `bgmaker.obs_table = my_obs_table` or giving it in the `__init__` already.

But that can also be done later ... the most important thing is that you get this example working with the new class (and get the same output):
https://github.com/gammapy/gammapy/blob/master/gammapy/scripts/tests/test_background_model.py#L22

OK?

**Comment by JouvinLea** [comment]
yes I agree with you, I was wondering the same things.
Let have the test works for the class as it was implement in background_model and then we will see how to improve it!

**Comment by JouvinLea** [comment]
By bgmaker.save() you mean to write the rate in a fits file? Because this is already done in make_model(). Perhaps you prefer to put the writing in an other method save() ?

**Comment by cdeil** [comment]
Yes, I think in principle it's nicer if the user has control over when / what gets saved to disk.
So there would be no or less saving to disk from the intermediate steps than with the command line, results would get stored as `bgmaker` data members.

There could also be some more control like:
```
bgmaker.obs_table.write('... exactly where I like ...')
bgmaker.model.rate.write('... exactly where I like ...')
```
and most users would use the default:
```
bgmaker.save(outdir='where I like', debug=True)
```
which saves the main result or also debug intermediate results (like counts, exposure cubes), to a given output folder.

But if you prefer to handle it some other way, that's fine with me, too.

**Comment by cdeil** [comment]
You could also have a look at this class as an example, which is of similar purpose (run some algorithm, save some output):
https://gammapy.readthedocs.org/en/latest/api/gammapy.background.IterativeKernelBackgroundEstimator.html

I'm not saying that one is very well done or you should follow the same pattern ... just an example to look at before you implement things here in a way you like.

**Comment by JouvinLea** [comment]
Hi,
I implemented the new class with the test that works.
As you said yesterday, I am not sure about the writing in the select_observation() and group_observation() methods?
Maybe we can close this PR is there is no conflict and after your suggestions? We can then improve it if we think this is needed later?

File: gammapy/background/off_data_background_maker.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+            if modeltype == "3D":
+                model = CubeBackgroundModel.define_cube_binning(obs_table_group, method='default')
+                model.fill_obs(obs_table_group, self.data_store)
+                model.smooth()
+                model.compute_rate()
+                self.models3D.append(model)
+
+            elif modeltype == "2D":
+                ebounds = EnergyBounds.equal_log_spacing(0.1, 100, 100, 'TeV')
+                offset = sqrt_space(start=0, stop=2.5, num=100) * u.deg
```
Comment (JouvinLea) [inline gammapy/background/off_data_background_maker.py]: I changed to a sqrt distribution.
Maybe we have to give the energy and offset as parameters no? This is weird to define them here?

File: gammapy/background/__init__.py [hunk 2]
```diff
@@ -10,3 +10,4 @@
 from .template import *
 from .kernel import *
 from .models import *
+from off_data_background_maker import *
```
Comment (cdeil) [inline gammapy/background/__init__.py]: You need to change to this to make it work on Python 3 also:
```python
from .off_data_background_maker import *
```

File: gammapy/background/off_data_background_maker.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    'OffDataBackgroundMaker',
+]
+
+log = logging.getLogger(__name__)
+
+
+class OffDataBackgroundMaker(object):
+    def __init__(self, data_store, outdir=None, run_list=None, obs_table_grouped_filename=None,
+                 group_table_filename=None):
+        """OffDataBackgroundMaker class.
```
Comment (cdeil) [inline gammapy/background/off_data_background_maker.py]: We always put the docstring for a class at the class level, and no docstring for `__init__`.
I.e. put this docstring here:
```
class MyClass(object):
    """Docstring here.
    """"
    def __init__():
         # no docstring here
```

File: gammapy/data/data_store.py [hunk 4]
```diff
@@ -234,6 +234,27 @@ def load_all(self, filetype):
             data_lists.append(data_list)
         return data_lists
 
+    def load_some_obs(self, list_obs, filetype):
```
Comment (cdeil) [inline gammapy/data/data_store.py]: Maybe rename to `load_many` for now?

Comment (cdeil) [inline gammapy/data/data_store.py]: Maybe rename to `load_many` for now?

File: gammapy/background/off_data_background_maker.py [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+
+        Parameters
+        ----------
+        data_store : `~gammapy.data.DataStore`
+            Data for the background model
+        run_list : str
+            filename where is store the OFF run list
+        outdir : str
+            directory where will go the output
+        obs_table_grouped_filename : str
```
Comment (cdeil) [inline gammapy/background/off_data_background_maker.py]: Is there value in having `obs_table_grouped_filename` and `group_table_filename` as parameters here?
Why not just store the default in `__init__`, and if the user really wants to customize those,
then they can do this:
```
maker = OffDataBackgroundMaker()
maker. obs_table_grouped_filename = 'my_custom_name'
```

File: gammapy/background/off_data_background_maker.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+        filename = self.obs_table_grouped_filename
+        log.info('Writing {}'.format(filename))
+        obs_table.write(str(filename), format='ascii.ecsv')
+
+        filename = self.group_table_filename
+        log.info('Writing {}'.format(filename))
+        obs_groups.obs_groups_table.write(str(filename), format='ascii.ecsv')
+        self.ntot_group = obs_groups.n_groups
+
+    def make_model(self, modeltype, obs_table=None, excluded_sources=None):
```
Comment (cdeil) [inline gammapy/background/off_data_background_maker.py]: I would remove the `obs_table` and `excluded_sources` parameters here and add those earlier, usually in `__init__`.
The basic pattern of usage of a maker would then be:
```
maker = Maker( ... parameters, inputs, configuration ...)
maker.run_step1()
maker.run_step2( option2_parameters )
```
I.e. the individual steps would only take a few parameters (if any) ... basically the algorithm is set up in `__init__`.

@JouvinLea - What do you think?

Comment (JouvinLea) [inline gammapy/background/off_data_background_maker.py]: Not sure to understand this last point…
For the fill_obs method we need the excluded_sources table! You want to directly define the obs_table and the excluded sources as member of the class?

File: gammapy/background/off_data_background_maker.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+        if not run_list:
+            self.run_list = "run.lis"
+        else:
+            self.run_list = run_list
+
+        if not outdir:
+            self.outdir = "out"
+        else:
+            self.outdir = outdir
+        self.obs_table = obs_table
```
Comment (JouvinLea) [inline gammapy/background/off_data_background_maker.py]: I added this two parameters as class members, was it what you think about?

**Pull request merged**